### PR TITLE
Notifiers with retry

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added retrial mechanism for the output notifiers

### CODE REVIEW

File: executors/notifiersHandler.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	if len(messages) == 0 {
+		return
+	}
+
+	for _, notifier := range handler.notifiers {
+		err := notifier.OutputMessages(messages...)
+		if err != nil {
+			uniqueNotifierNameStr := uniqueNotifierName(notifier)
+			log.Error("error sending notification", "notifier", uniqueNotifierNameStr, "error", err)
+			notifiersThatErrored[uniqueNotifierNameStr] = notifier
```
Comment (andreibancioiu) [inline executors/notifiersHandler.go:66]: Nice solution :+1:

File: executors/notifiersHandler.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	return fmt.Errorf("%w: num notifiers with problems: %d", errNotificationsSendingProblems, len(notifiersThatErrored))
+}
+
+func (handler *notifiersHandler) notifyInRetryMode(retryNumber uint32, notifiersThatErrored map[string]OutputNotifier, messages []core.OutputMessage) {
+	for identifier, notifier := range notifiersThatErrored {
+		err := notifier.OutputMessages(messages...)
+		if err == nil {
+			delete(notifiersThatErrored, identifier)
```
Comment (andreibancioiu) [inline executors/notifiersHandler.go:93]: :+1:

File: executors/notifiersHandler.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+// NotifyWithRetry will try to send the messages to the inner list of notifiers. If one or more notifiers errored, it will retry
+// for the maximum number of retries provided. There is a delay between the retries, to maximize the chances of recovery
+func (handler *notifiersHandler) NotifyWithRetry(caller string, messages ...core.OutputMessage) error {
+	log.Debug("notifiersHandler.NotifyWithRetry",
+		"executor", caller, "num messages", len(messages), "num retries", handler.numRetries,
+		"num notifiers", len(handler.notifiers))
+
+	notifiersThatErrored := make(map[string]OutputNotifier)
+	handler.handleFirstSend(notifiersThatErrored, messages)
+	return handler.handleRetries(notifiersThatErrored, messages)
```
Comment (andreibancioiu) [inline executors/notifiersHandler.go:53]: If one notifier is permanently failing it does not block other notifications from happening :+1:

**Review by andreibancioiu — APPROVED** [review]

File: executors/statusHandler.go [hunk 4]
```diff
@@ -11,26 +11,26 @@ import (
...
+	notifiersHandler OutputNotifiersHandler
 }
 
 // NewStatusHandler creates a new instance of type statusHandler
-func NewStatusHandler(name string, notifiers []OutputNotifier) (*statusHandler, error) {
-	for index, notifier := range notifiers {
-		if check.IfNil(notifier) {
-			return nil, fmt.Errorf("%w at index %d", errNilOutputNotifier, index)
-		}
+func NewStatusHandler(name string, notifiersHandler OutputNotifiersHandler) (*statusHandler, error) {
```
Comment (miiu96) [inline executors/statusHandler.go:25]: add a check for the `name` parameter to not be empty ?

**Review by miiu96 — APPROVED** [review]

File: executors/notifiersHandler.go (hunk 3 above)
Comment (iulianpascalau) [reply executors/notifiersHandler.go:53]: yes, indeed, it will signal problems in the logs

File: executors/statusHandler.go (hunk 4 above)
Comment (iulianpascalau) [reply executors/statusHandler.go:25]: it can be empty, it is mostly used on logs

**Pull request merged**